File: 2025/05 May/04 Smallest distinct window/main.cpp
// Link: https://www.geeksforgeeks.org/problems/smallest-distant-window3132/1

#include <iostream>
#include <unordered_set>
#include <map>
using namespace std;

class Solution
{
public:
    int findSubString(string &str)
    {
        unordered_set<char> st;
        for (auto &i : str)
            st.insert(i);
        int count = st.size();
        map<int, int> mp;
        int n = str.length();
        int ans = n;
        int i = 0, j = 0;
        while (j < n)
        {
            mp[str[j]]++;
            while (mp.size() == count)
            {
                ans = min(ans, j - i + 1);
                mp[str[i]]--;
                if (mp[str[i]] == 0)
                    mp.erase(str[i]);
                i++;
            }
            j++;
        }
        while (mp.size() == count)
        {
            ans = min(ans, j - i + 1);
            mp[str[i]]--;
            if (mp[str[i]] == 0)
                mp.erase(str[i]);
            i++;
        }
        return ans;
    }
};

int main()
{
    Solution obj;
    string str = "aabcbcdbca";
    cout << "The length of the smallest distinct window is: " << obj.findSubString(str) << endl;

    str = "aaab";
    cout << "The length of the smallest distinct window is: " << obj.findSubString(str) << endl;

    str = "geeksforgeeks";
    cout << "The length of the smallest distinct window is: " << obj.findSubString(str) << endl;
    return 0;
}

File: 2025/05 May/04 Smallest distinct window/main.py
# Link: https://www.geeksforgeeks.org/problems/smallest-distant-window3132/1


class Solution:
    def findSubString(self, str):
        st = set(str)
        count = len(st)
        mp = {}
        n = len(str)
        ans = n
        i = 0
        j = 0

        while j < n:
            mp[str[j]] = mp.get(str[j], 0) + 1
            while len(mp) == count:
                ans = min(ans, j - i + 1)
                mp[str[i]] -= 1
                if mp[str[i]] == 0:
                    del mp[str[i]]
                i += 1
            j += 1

        while len(mp) == count:
            ans = min(ans, j - i + 1)
            mp[str[i]] -= 1
            if mp[str[i]] == 0:
                del mp[str[i]]
            i += 1

        return ans


if __name__ == "__main__":
    obj = Solution()
    str = "aabcbcdbca"
    print("The length of the smallest distinct window is:", obj.findSubString(str))

    str = "aaab"
    print("The length of the smallest distinct window is:", obj.findSubString(str))

    str = "geeksforgeeks"
    print("The length of the smallest distinct window is:", obj.findSubString(str))
